File: leetcode/cpp/PassedProblems/264-ugly-number-ii.cpp
/*
 * 264-ugly-number-ii.cpp
 * Write a program to find the n-th ugly number.

Ugly numbers are positive numbers whose prime factors only include 2, 3, 5.
For example, 1, 2, 3, 4, 5, 6, 8, 9, 10, 12 is the sequence of the first 10 ugly numbers.

Note that 1 is typically treated as an ugly number.
 *  Created on: Oct 4, 2015
 *      Author: hanzhe
 */

class Solution {
public:
	int myMin(int a, int b, int c){
		int t = a > b ? b : a;
		return t > c ? c : t;
	}
    int nthUglyNumber(int n) {
    	int * p = new int[n+1];
    	p[0] = 1;
    	int i2 = 0, i3 = 0, i5 = 0;
		int f2 = p[i2] * 2, f3 = p[i3] * 3, f5 = p[i5] * 5;
    	for(int i = 1 ; i < n; i ++){
    		int min = myMin(f2, f3, f5);
    		p[i] = min;
    		if(min == f2)
    			f2 = 2 * p[++i2];
    		if(min == f3)
    			f3 = 3 * p[++i3];
    		if(min == f5)
    			f5 = 5 * p[++i5];
    	}
    	return p[n-1];
    }
};

File: leetcode/python/264-ugly-number-ii.py
#!/usr/bin/env python
# encoding: utf-8


"""

Write a program to find the n-th ugly number.

Ugly numbers are positive numbers whose prime factors only include 2, 3, 5. For example, 1, 2, 3, 4, 5, 6, 8, 9, 10, 12 is the sequence of the first 10 ugly numbers.

Note that 1 is typically treated as an ugly number.
"""

# 2次过，速度慢
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1: return 1
        bkt = [[2],[3],[5]]
        a = [2,3,5]
        last = 1
        while n > 1:
            for i in range(0,3):
                while bkt[i][0] <= last: bkt[i].pop(0)
            mi, mv = 0, bkt[0][0]
            for i in range(1,3):
                if bkt[i][0] < mv:
                    mi, mv = i, bkt[i][0]
            last = mv
            bkt[mi].pop(0)
            for i in range(0,3):
                bkt[i].append(last * a[i])
            n -= 1
        return last



